src/framework/audio/internal/audiobuffer.cpp

```diff
@@ -0,0 +1,128 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2020 MuseScore BVBA and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#include "audiobuffer.h"
+#include <cstring>
+#include "log.h"
+
+using namespace mu::audio;
+
+AudioBuffer::AudioBuffer(unsigned int streamsPerSample, unsigned int size)
+    : m_streamsPerSample(streamsPerSample)
+{
+    m_data.resize(size * m_streamsPerSample, 0.f);
+}
+
+void AudioBuffer::setSource(std::shared_ptr<IAudioSource> source)
+{
+    m_source = source;
```

Comment:
The thread-unsafe code.
This method can be called (and is called) on the main thread, and the `m_source` access from the worker thread.